kimchi/src/circuits/constraints.rs

```diff
@@ -654,7 +658,9 @@ impl<F: FftField + SquareRootField> ConstraintSystem<F> {
         // ------
 
         let lookup_constraint_system =
-            LookupConstraintSystem::create(&gates, lookup_tables, &domain).ok()?;
+            LookupConstraintSystem::create(&gates, lookup_tables, &domain)
```

Comment:
@mimoo  `LookupConstraintSystem::create()` returns a somewhat weird combination of `Result<Option<Self>, LookupError>` - not sure if it would be better to return a `Result<LookupConstraintSystem, LookupError>` instead

Reply:
yeah that would be better

Reply:
actually, sorry, that looks normal to me (albeit a bit weird). Since we don't know at this point if lookups will be used. The function analyzes the circuit and returns `None` if no gates are used that make use of lookups.

Reply:
so here just doing `ok()?` is not going to work because you have two different error types. Not sure what's the best way forward:

1. create a `SetupError` error type?
2. create an enum `ProverError::Lookup(LookupError)` and use `map_err` to convert to that
3. something else

I would personally go for 1. and convert the error of `LookupConstraintSystem::create` to such an error type. Others might have some opinions (cc @mrmr1993 @querolita @jspada)

Reply:
> actually, sorry, that looks normal to me (albeit a bit weird). Since we don't know at this point if lookups will be used. The function analyzes the circuit and returns `None` if no gates are used that make use of lookups.

ye thats right, I saw that after I was looking into as well

Reply:
btw I just hit that thing, and it's going to be really useful to return an error here instead of returning `None` (which doesn't tell you where the error happened)

Reply:
> btw I just hit that unwrap. It's going to be really useful to return an error here instead of returning None (which doesn't tell you where the error happened)

 so you think its worth it re-writing the `Result<Option<Self>, LookupError>` from `LookupConstraintSystem::create()` to just return a normal `Result<LookupConstraintSystem, LookupError>` instead of the Option `None` then? Can do it that way if that's how you want to deal with the nested `Result<Option, _>`

Reply:
oh no, the `Option` is important to keep, I meant more generally the constraint system (not the lookup constraint system) doesn't return an error and we end up unwrapping an option if we reach that case (which doesn't tell you why it didn't work)